### PR TITLE
Boos 4345 gnn+llm graph reasoning ft

### PR DESCRIPTION
**Pull request by Johnno1011**
This pr includes:
- Fine tuning and pretraining of LLM and GNN respectively using crossentropy loss.
- Grounding the GLLM responses to generic graph related questions with responses from openai models on a vanilla text prompt.
- Tweaks to LORA FT config and tracking of loss and saving of model checkpoints during training.

### CODE REVIEW

File: src/primaite/agents/aegis/data.py [hunk 1]
```diff
@@ -9,9 +9,12 @@ class GLLMDataset(Dataset):
     def __init__(self, graphs, questions, gt_answers, llm: LLM):
         self.graphs = graphs
         self.questions = questions
-        self.gt_answers = gt_answers
+        self.gt_answers = [
+            answer + " " + llm.tokenizer.eos_token for answer in gt_answers
```
Comment (HarveyAyling) [inline src/primaite/agents/aegis/data.py]: Maybe remove the additional space " " as it is no longer needed.

File: src/primaite/agents/aegis/data.py [hunk 2]
```diff
@@ -22,21 +25,24 @@ def __getitem__(self, idx):
...
+        inputs = self.llm.get_input_embeddings(prompts=gt_answers, grad=False)
+        hidden_states = self.llm.get_output_embeddings(inputs_embeds=inputs["inputs_embeds"], grad=False).to("cpu")
         return hidden_states
 
+    def _get_target_logprobs_and_tokens(self, gt_answers: List[str]):
+        # Construct mock logprobs for gt repsonse
+        gt_tokens = self.llm.tokenizer.batch_encode_plus(gt_answers, return_tensors="pt", padding=True)["input_ids"]
+        batch_size = gt_tokens.shape[0]
+        gt_probs = torch.zeros((batch_size, gt_tokens.shape[1], len(self.llm.tokenizer)), dtype=torch.float32)
+        for token_idx in range(gt_tokens.shape[1]):
```
Comment (HarveyAyling) [inline src/primaite/agents/aegis/data.py]: This loop can be done in one operation using like a torch scatter, but its probably not that performance detrimental, im just being picky, and torch scatter is kinda complicated lol

File: src/primaite/agents/aegis/gllm.py [hunk 3]
```diff
@@ -89,10 +96,62 @@ def get_cosine_embeddings_loss(self, openai_hidden_states, gllm_hidden_states):
...
+        padding_logit = torch.zeros([len(self.llm.tokenizer)])
+        padding_logit[self.llm.tokenizer.pad_token_id] = (
+            100  # Mock pad token of gt to 100% prob (it gets ignored anyway because of the mask)
+        )
+        padding_token = torch.tensor([self.llm.tokenizer.pad_token_id])  # Set padding token tensor
+
+        # Pad the shortest set of tensors if required
+        batch_size = gllm_probs.shape[0]
+        if gt_probs.shape[1] > gllm_probs.shape[1]:
+            # gt sequences are longer - pad gllm tokens
```
Comment (HarveyAyling) [inline src/primaite/agents/aegis/gllm.py]: Please re-write the comments now this is changed. As this is truncating the ground truth responses, no longer padding :)

Comment (HarveyAyling) [reply src/primaite/agents/aegis/gllm.py]: Also you can get rid of `n_pads` from this part directly below

File: src/primaite/agents/aegis/gllm.py [hunk 4]
```diff
@@ -89,10 +96,62 @@ def get_cosine_embeddings_loss(self, openai_hidden_states, gllm_hidden_states):
...
+            padding_logits = padding_logit.repeat(batch_size, n_pads, 1)
+
+            gt_probs = torch.cat([gt_probs, padding_logits], dim=1)
+
+            padding_tokens = padding_token.repeat(batch_size, n_pads)
+            gt_tokens = torch.cat([gt_tokens, padding_tokens], dim=1)
+
+        # Calculate the cross entropy loss - transpose the 1 and 2 dimensions because we want a loss value per token not per vocab
+        loss = criterion(gllm_probs.transpose(1, 2), gt_probs.transpose(1, 2))
+        loss = loss.mean()
```
Comment (HarveyAyling) [inline src/primaite/agents/aegis/gllm.py]: Again very pedantic, but might as well just remove the `reduction="none"` from `criterion = torch.nn.CrossEntropyLoss(reduction="none")` and then can also remove `loss = loss.mean()`

File: src/primaite/agents/aegis/gllm.py [hunk 5]
```diff
@@ -89,10 +96,62 @@ def get_cosine_embeddings_loss(self, openai_hidden_states, gllm_hidden_states):
...
+
+
+def get_n_trainable_llm_parameters(model):
+    trainable_model_params = 0
+    all_model_params = 0
+    for _, param in model.named_parameters():
+        all_model_params += param.numel()
+        if param.requires_grad:
+            trainable_model_params += param.numel()
+    return f"Trainable LLM parameters: {trainable_model_params}\nAll LLM parameters: {all_model_params}\nPercentage of trainable LLM parameters: {100 * trainable_model_params / all_model_params:.2f}%"
```
Comment (HarveyAyling) [inline src/primaite/agents/aegis/gllm.py]: I am a Python nerd :) so please make this f-string nice lol 
If you add like `{trainable_model_params:_}` it will print large numbers like 9_000_000
And for percentages you can just `{trainable_model_params / all_model_params:.2%}`

File: src/primaite/agents/aegis/gllm.py [hunk 6]
```diff
@@ -113,6 +181,16 @@ def train_loop(model: GLLM, dataloader: DataLoader, network_desc: str, n_epochs:
 
             model.optimizer.step()
 
-        print(f"Epoch {epoch} Loss:", loss.item(), gllm_responses[0].replace("\n", "//n"))
-
+        if epoch % checkpoint_every == 0:
```
Comment (HarveyAyling) [inline src/primaite/agents/aegis/gllm.py:181]: Loving the checkpointing very nice 👍

**Review by HarveyAyling — COMMENTED** [review]
Looking goood, just a few pedantic changed from a Python Pytorch nerd 
💪 💪

File: src/primaite/agents/aegis/gllm.py (hunk 5 above)
Comment (Johnno1011) [reply src/primaite/agents/aegis/gllm.py]: Mate take this up with deeplearning.ai cos I robbed this code from their course generative ai with LLMs xD

File: src/primaite/agents/aegis/data.py (hunk 1 above)
Comment (Johnno1011) [reply src/primaite/agents/aegis/data.py]: thakns

File: src/primaite/agents/aegis/gllm.py (hunk 6 above)
Comment (Johnno1011) [reply src/primaite/agents/aegis/gllm.py:181]: If you actually ran it you'd know that there was a bug <3

File: src/primaite/agents/aegis/gllm.py (hunk 4 above)
Comment (Johnno1011) [reply src/primaite/agents/aegis/gllm.py]: Fair

File: src/primaite/agents/aegis/gllm.py (hunk 5 above)
Comment (HarveyAyling) [reply src/primaite/agents/aegis/gllm.py]: Shocking from them

File: src/primaite/agents/aegis/gllm.py (hunk 6 above)
Comment (HarveyAyling) [reply src/primaite/agents/aegis/gllm.py:181]: 💀

File: src/primaite/agents/aegis/gllm.py (hunk 5 above)
Comment (Johnno1011) [reply src/primaite/agents/aegis/gllm.py]: bruvvah I didn't even know about :_ or :, wowowwww

**Comment by Johnno1011** [comment]
just gonna wait for a nifets review

File: src/primaite/agents/aegis/gllm.py [hunk 7]
```diff
@@ -101,10 +157,19 @@ def train_loop(model: GLLM, dataloader: DataLoader, network_desc: str, n_epochs:
...
+            if loss_fn == "cosine":
+                loss = model.get_cosine_embeddings_loss(batch["gt_hidden_states"], gllm_hidden_states)
+            elif loss_fn == "crossentropy":
+
+                loss = model.get_crossentropy_loss(
+                    gt_probs=batch["gt_logprobs"],
+                    gt_tokens=batch["gt_tokens"],
+                    gllm_probs=gllm_logprobs,
+                    gllm_tokens=gllm_tokens,
+                )
```
Comment (nifets-ai) [inline src/primaite/agents/aegis/gllm.py:172]: does it make more sense to initialise this loss before the for loop, since it's global across the batches anyway

Comment (nifets-ai) [reply src/primaite/agents/aegis/gllm.py:172]: dont htink it makes a difference, jsut a style thing

Comment (nifets-ai) [reply src/primaite/agents/aegis/gllm.py:172]: nvm me
